Simpler ordering when printing certain additions of constants and negative expressions

nickovs opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->

Fixes #15982

#### Brief description of what is fixed or changed

This PR changes the print order of two-argument `Add` objects when one argument is positive `Rational` and the other is a `Mul` multiplying a negative `Rational` and another expression.

The purpose is to provide a more natural representation for values such as `1 - x` or `2 - sin(t)/2`. Currently these print as `-x + 1` and `-sin(t)/2 + 2` respectively, which are forms that would rarely be written by a human unless trying to make a specific point.

#### Other comments

The vast majority of this PR is updating the examples embedded in the source so that `doctest` passes. The only actual code change is in: https://github.com/sympy/sympy/commit/c52ebc834a89275433f7cbd3b596f66f6930a59c

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

* printing
  * changed the print order of additions of form `k - expr` (used to print as `-expr + k`)

<!-- END RELEASE NOTES -->

nickovs committed:
Prettier printing for Add(+Rational, Mul(-Rational, expr))

nickovs committed:
Updated examples so that doctests pass with the updated print order.

nickovs commented:
I should note that this is my first PR to this project, and I am quite new to the sympy source code. As such I am very open to criticism of this and happy to make changes as needed.

I started a discussion of this topic in issue #15982. If there needs to be more discussion of the use cases or the correct solution to this problem, rather than this code in particular, it might make sense to do it there.

oscarbenjamin commented:
How does the example in #15931 come out with this PR (just out of interest). That issue looked like another example where there was room for improvement but might be orthogonal to this...

oscarbenjamin commented:
There are a bunch of test fails. Check the details link by "continuous-integration/travis..." below.

nickovs commented:
@oscarbenjamin I just spotted that. I will investigate.

asmeurer commented:
I didn't notice there was a PR already. Reposting this here from the issue.

This ignores the order if the order argument is passed to the printer. Changing the default ordering is fine, but `order='lex'` and so on should continue to give that exact order for polynomials.

It might be better to make this change in `as_ordered_terms`. Unfortunately, that may result in even more changes, as that affects the sort key ordering as well as the printing ordering (it may even unearth some bugs, see for instance #13059).

nickovs committed:
Moved code from printing/printer.py to core/expr.py

Fixed numerous test cases to reflect both the updated string representation
and the impact that that has on the sort order for various results.

nickovs committed:
Deleted annoying trailing whitespace in test case.

nickovs commented:
I've updated the code to implement this in `as_ordered_terms()` and obey the `order` parameter. The relevant code is in https://github.com/sympy/sympy/pull/15985/commits/a3b984d363daac004449f5230f5c0dbbb1da1fc1#diff-f13f7d02dee32a4be2f221ab3feb3458 

The rest of the patch is fixing up test cases that broke, in many case because with the new order of printing the sort order for various result sets has changed.

oscarbenjamin commented on sympy/solvers/ode.py:
```diff
@@ -4764,13 +4764,13 @@ def ode_separable_reduced(eq, func, order, match):
-                     x                           x
+                   ___________            ___________
+                  /     2                /     2
+            1 - \/  C1*x  + 1          \/  C1*x  + 1  + 1
```
This is one of the only cases where I preferred the previous output. It loses the symmetry between 1+a,1-a by becoming 1-a,a+1.

nickovs replied:
In this case, and many others with roots, the best option would be to have `1+sqrt(x)` and `1-sqrt(x)`. In the light of the comment that you raise below, it's going to be hard to please all of the cases all of the time. As I went through all of the fix-ups for the test cases I did start to wonder whether there is a case for a contextual formatter when formatting sets of results, which aims to present the symmetries, but that's a whole different can of worms.

Still, I think based on all of the patches to the tests that we needed to do this, that on balance this makes things look better a great deal more often than it makes them worse.

oscarbenjamin commented on sympy/solvers/tests/test_solvers.py:
```diff
@@ -1624,8 +1624,8 @@ def test_rewrite_trig():
-    assert solve(2*sin(x) - cos(x), x) == [-2*atan(2 + sqrt(5)),
-                                           -2*atan(-sqrt(5) + 2)]
+    assert solve(2*sin(x) - cos(x), x) == [-2*atan(2 - sqrt(5)),
+                                           -2*atan(2 + sqrt(5))]
```
This one is the reverse of the above. We had a+b,-b+a and now we have a-b,a+b which is nicer.

oscarbenjamin commented:
I almost always prefer the new output looking through the doctest changes.

I'm still unsure about special casing two terms though. Do we not still get the same problem with 3 terms?
```julia
In [18]: 1-x-x**2                                                                                                                                             
Out[18]: 
   2        
- x  - x + 1
```

nickovs commented:
> I'm still unsure about special casing two terms though. Do we not still get the same problem with 3 terms?

We could certainly do this for longer polynomial too but I find `-x² -x + 1` is less jarring than `-x + 1` and I also think that if we start reversing longer polynomials the number of cases where you loose symmetry will go up.

oscargus commented:
I think this is a clear step forward. There will, as noted, always be some cases that can look better, but this fixes one of the more obvious ones. I'm in favor of merging.

smichr commented:
A lot of people will smile after this...or at least frown less. Thanks.

oscarbenjamin commented:
Yes thanks @nickovs!

oscarbenjamin commented:
I just found this:
```julia
In [14]: integrate(2 * sqrt(-x**2 + 1)*(4*sqrt(2)*x**2 - sqrt(3)), (x, -1, +1))                                                   
Out[14]: -√3⋅π + √2⋅π
```
Haven't looked into the details yet but do you have any idea why this happens @nickovs:
```julia
In [18]: sqrt(2)-1                                                                                                                
Out[18]: -1 + √2
```

nickovs commented:
Both cases are the result of the existing code being very conservative as to what it will or won't reorder, and neither of these match the pattern.

In the first case the `srepr()` for the result is `Add(Mul(Integer(-1), Pow(Integer(3), Rational(1, 2)), pi), Mul(Pow(Integer(2), Rational(1, 2)), pi))`, so both sides of the addition are `Mul` nodes and the current code only reorders when one side is a number and the other side a multiplication.

In the other case the srepr() is `Add(Integer(-1), Pow(Integer(2), Rational(1, 2)))`, and again this does not match the addition between a number and a multiplication.

There are probably many more cases that don't format as well as one would like. As it stands all of these should print exactly as they did before. Can I suggest that you collect together all the examples that you can find of things that do not print as you would like and open another issue where we can discuss how to better handle this. I'm happy to write more general code than the very narrow solution that was merged here but I think if I'm going to rewrite this we should think through all the cases that we care about before I make a new PR.

oscarbenjamin commented:
Thanks for getting back @nickovs. I've opened #16090
